src/input/json_to_storage.cc

```diff
@@ -336,21 +336,39 @@ StorageBase * JSONToStorage::make_storage(JSONPath &p, const Type::Record *recor
 {
     if (p.head()->type() == json_spirit::obj_type) {
     	const json_spirit::mObject & j_map = p.head()->get_obj();
-    	std::set<string> keys_to_processed;
+    	std::set<string> keys_to_process;
         json_spirit::mObject::const_iterator map_it;
         std::set<string>::iterator set_it;
 
         for( map_it = j_map.begin(); map_it != j_map.end(); ++map_it) {
-           keys_to_processed.insert(map_it->first);
+           keys_to_process.insert(map_it->first);
+        }
+
+        Type::Record::KeyIter key_it;
+        if ( record->has_key_iterator("TYPE", key_it) && record->auto_conversion_key_iter() != record->end() ) {
+            JSONPath type_path(p);
+            if (type_path.down( "TYPE" ) != NULL) {
+                try {
+                	if ( type_path.head()->get_str() != record->type_name() ) {
+                		xprintf(UsrErr, "Invalid value of TYPE key of record %s.", record->type_name().c_str());
```

Comment:
Resolution:
* set default value of TYPE key to the Record name v has_obligatory_type_key
* Record::finish - check that autoconvertible key is of type Record, in such a case check that TYPE key is declared. Optionally introduce flag: "is_autoconvertible_from_record" to simplify similar test in JSONToStorage.

	type_id(*data_->keys[data_->autoconvertible_key_idx_].type_) == Record